src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -70,45 +74,57 @@ public void get(String key, final OnGetCompleteCallback callback) {
             final AsynchronousFileChannel afc =
                     AsynchronousFileChannel.open(path, StandardOpenOption.READ);
             channel = afc;
-            int headerSize = entry.getHeaderSize();
-            final int size = (int) file.length() - headerSize;
+            final int size = (int) file.length();
             final ByteBuffer buffer = ByteBuffer.allocate(size);
             afc.read(
                     /* destination= */ buffer,
-                    /* position= */ headerSize,
+                    /* position= */ 0,
                     /* attachment= */ null,
                     new CompletionHandler<Integer, Void>() {
                         @Override
                         public void completed(Integer result, Void v) {
                             closeChannel(afc, "completed read");
-                            // if the file size changes, return null
                             if (size != result) {
                                 VolleyLog.e(
                                         "File changed while reading: %s", file.getAbsolutePath());
-                                callback.onGetComplete(null);
+                                removeDeleteAndCall(key, callback, file);
                                 return;
                             }
-                            byte[] data = buffer.array();
-                            callback.onGetComplete(entry.toCacheEntry(data));
+                            buffer.flip();
+                            CacheHeader entryOnDisk = CacheHeader.readHeader(buffer);
+                            if (entryOnDisk == null) {
+                                // BufferUnderflowException was thrown while reading header
+                                removeDeleteAndCall(key, callback, file);
+                            } else if (!TextUtils.equals(key, entryOnDisk.key)) {
+                                // File shared by two keys and holds data for a different entry!
+                                VolleyLog.d(
+                                        "%s: key=%s, found=%s",
+                                        file.getAbsolutePath(), key, entryOnDisk.key);
+                                removeDeleteAndCall(key, callback, file);
+                            } else {
+                                byte[] data = new byte[buffer.remaining()];
+                                buffer.get(data);
+                                callback.onGetComplete(entry.toCacheEntry(data));
+                            }
                         }
 
                         @Override
                         public void failed(Throwable exc, Void ignore) {
+                            closeChannel(afc, "completed read");
```

Comment:
not sure why this was reverted from "failed" to "completed"

Reply:
I don't know how I did that. Just changed it back.